runtime/src/accounts_index.rs

```diff
@@ -1179,30 +1179,49 @@ impl<T: 'static + Clone + IsCached + ZeroLamport> AccountsIndex<T> {
         }
     }
 
-    pub(crate) fn get_account_maps_write_lock(&self) -> AccountMapsWriteLock<T> {
+    fn get_account_maps_write_lock(&self) -> AccountMapsWriteLock<T> {
         self.account_maps.write().unwrap()
     }
 
-    // Same functionally to upsert, but doesn't take the read lock
-    // initially on the accounts_map
+    // Same functionally to upsert, but:
+    // 1. operates on a batch of items
+    // 2. holds the write lock for the duration of adding the items
     // Can save time when inserting lots of new keys.
     // But, does NOT update secondary index
+    // This is designed to be called at startup time.
+    #[allow(clippy::needless_collect)]
     pub(crate) fn insert_new_if_missing_into_primary_index(
         &self,
         slot: Slot,
-        pubkey: &Pubkey,
-        account_info: T,
-        reclaims: &mut SlotList<T>,
-        w_account_maps: &mut AccountMapsWriteLock<T>,
+        items: Vec<(&Pubkey, T)>,
     ) {
-        let account_entry =
-            self.insert_new_entry_if_missing(pubkey, slot, &account_info, Some(w_account_maps));
-        if account_info.is_zero_lamport() {
-            self.zero_lamport_pubkeys.insert(*pubkey);
-        }
-        if let Some(mut w_account_entry) = account_entry {
-            w_account_entry.update(slot, account_info, reclaims);
-        }
+        let potentially_new_items = items
+            .iter()
+            .map(|(_pubkey, account_info)| {
+                // this value is equivalent to what update() below would have created if we inserted a new item
+                WriteAccountMapEntry::new_entry_after_update(slot, account_info)
+            })
+            .collect::<Vec<_>>(); // collect here so we have created all data prior to obtaining lock
+
+        let mut _reclaims = SlotList::new();
+
+        let mut w_account_maps = self.get_account_maps_write_lock();
+        items
+            .into_iter()
+            .zip(potentially_new_items.into_iter())
+            .for_each(|((pubkey, account_info), new_item)| {
+                let account_entry = self.insert_new_entry_if_missing_with_lock(
+                    pubkey,
+                    &mut w_account_maps,
+                    new_item,
+                );
+                if account_info.is_zero_lamport() {
```

Comment:
I left this where it was in this pr. It doesn't appear to make a big difference in metrics.